### PR TITLE
v3: implement new symbol rules for ORDER BY

### PR DESCRIPTION
**Pull request by sougou**
This change addresses #2254. The ORDER BY clause is allowed
to reference columns in the FROM clause that are not necessarily
present in the SELECT expressions.

### CODE REVIEW

**Comment by alainjobart** [comment]
LGTM

**Comment by sougou** [comment]
Thinking about this some more, passing the forOrderBy flag as parameter is not the correct design. This is because the search rules for the outer symbol tables should not be dictated by a subquery, but the clause in which it appears. For example, in `select ... having id in (select ...)`, the inner subquery symbols have to follow the rules dictated by the `having` clause, and should not depend on the context in which they appear in the subquery.

TL;DR: The flag should actually be a state in the symbol table, and it will individually dictate how that symbol table performs the search. In reality, this will have no material impact for this feature, but will become relevant when we start to handle more constructs.

Update coming soon.
